core/src/test/scala/unit/kafka/utils/UtilsTest.scala

```diff
@@ -177,4 +181,33 @@ class UtilsTest extends JUnitSuite {
     assertEquals(clusterId.length, 22)
     assertTrue(clusterIdPattern.matcher(clusterId).matches())
   }
+
+  @Test
+  def testGetOrElseUpdateAtomically(): Unit = {
+    val count = 1000
+    val nThreads = 5
+    val createdCount = new AtomicInteger
+    val map = new ConcurrentHashMap[Int, AtomicInteger]().asScala
+    val executor = Executors.newFixedThreadPool(nThreads)
+    try {
+      for (i <- 1 to count) {
+        executor.submit(new Runnable() {
+          def run() {
+            CoreUtils.getOrElseUpdateAtomically(map, 0, {
```

Comment:
Out of curiosity, does this fail if you use the non atomic implementation?

Reply:
Yes, it does consistently fail on my laptop:
```
java.lang.AssertionError: expected:<1000> but was:<995>
```

Reply:
Perfect.